### DISCUSSION TITLE
Allowing TAs and professors to submit answers on behalf of students

### DISCUSSION THREAD

**Discussion by kitsook** [openwebwork/webwork2#1179] [General]
Hi,

We encountered some cases that students may not submit their answers for grading within the quiz time limit.  Thus, we tried to enable TAs and professors to submit answers for students.

We tried to set
```
$permissionLevels{record_answers_when_acting_as_student} = "ta";
$permissionLevels{record_set_version_answers_when_acting_as_student} = "ta";
```
but it doesn't seem to work.

It worked when we set
```
$permissionLevels{avoid_recording_answers} = "nobody";
```
and TAs and professors can click on "Grade Test" with successful results.

Anyone encountered similar issues? Any reported bugs on `record_answers_when_acting_as_student` and `record_set_version_answers_when_acting_as_student`?

**Comment by dlglin** [comment]
The two permissions you changed allow users to submit on behalf of other users, but this is still subject to other restrictions.  In particular it will not allow a ta or professor to submit answers on behalf of the student after the due date or when the student is out of attempts.

I believe that you will also have to change the following permissions:
```
        record_answers_after_open_date_without_attempts
        record_answers_after_due_date
        record_answers_after_answer_date
```

**Reply by kitsook** [reply]
Thanks for the suggestion.  I tried these permissions:
```
$permissionLevels{record_answers_when_acting_as_student} = "ta";
$permissionLevels{record_set_version_answers_when_acting_as_student} = "ta";
$permissionLevels{record_answers_after_open_date_without_attempts} = "ta";
$permissionLevels{record_answers_after_due_date} = "ta";
$permissionLevels{record_answers_after_answer_date} = "ta";
```

Instructors still can't submit on behalf of students.  Unless, `avoid_recording_answers` is also set to `admin` or `nobody`.  Possible that `avoid_recording_answers` has higher priority and limiting the effectiveness of other permissions?

**Reply by kitsook** [reply]
Seems that `avoid_recording_answers` does override other permissions.  Basically it is a battle between whether the user "can" record answers vs whether the user "want to" record answers.

https://github.com/openwebwork/webwork2/blob/03eca4cc1a6034d24cf28bd690f0f56ce14f284b/lib/WeBWorK/ContentGenerator/GatewayQuiz.pm#L1155-L1157

So whether webwork "will" record answers is based on if the user "can" record answers **and** if user "want to" record answers (let's forget the "must" part for now).

Setting these permissions will give instructors the "can" record answers ability:
```
$permissionLevels{record_answers_when_acting_as_student} = "ta";
$permissionLevels{record_set_version_answers_when_acting_as_student} = "ta";
```

However, the "want to" record answers setting is based on the permission `avoid_recording_answers`:
https://github.com/openwebwork/webwork2/blob/03eca4cc1a6034d24cf28bd690f0f56ce14f284b/lib/WeBWorK/ContentGenerator/GatewayQuiz.pm#L1111-L1112

That means `avoid_recording_answers` needs to be set as `admin` or `nobody` so that webwork knows instructors "want to" record answers if them clicked on the "Grade Test" button.

**Comment by drgrice1** [comment]
You should note that `$permissionLevels{avoid_recording_answers} = "nobody"` is the default setting.  If that is different on your server, then it was changed.

There really isn't this battle of the can, will, must, want hashes as you say.  The avoid_recording_answers permission is doing what it is intended to do.  It overrides the other permissions that you are talking about.
